### PR TITLE
[RFC] add contributor principles and merge policy

### PR DESCRIPTION
**Pull request by Stebalien**
cc @ipfs/go-team

This is a proposal, please discuss. If you don't think this answers questions like "can I merge this?", "can I change this?", etc., please say so.

Also, it would be *awesome* to get some examples here.

### CODE REVIEW

**Comment by Stebalien** [comment]
Also cc @momack2.

File: CONTRIBUTING_GO.md [hunk 1]
```diff
@@ -161,6 +164,27 @@ reasons why this is _the right thing to do_:
...
 Signed-off-by: Juan Benet <juan@ipfs.io>
 ```
+
+## Contributor Principles
+
+These principles are a living document and should be updated as-needed to better serve the go-ipfs community.
+
+In order of precedence:
+
+1. **Build the distributed web.** (may need a better way to capture this)
```
Comment (momack2) [inline CONTRIBUTING_GO.md]: aka "This is the goal - never compromise the goal"
```suggestion
1. **Build the distributed peer-to-peer web.** This is our objective and aim - everything we do is in service of this goal and we only compromise our other principles if needed to achieve this.
```

File: CONTRIBUTING_GO.md [hunk 2]
```diff
@@ -161,6 +164,27 @@ reasons why this is _the right thing to do_:
...
+2. **Don't break userspace.** This isn't "don't break protocols"; we do, in fact, occasionally break protocols (with care) when we know that users won't notice it.
```
Comment (momack2) [inline CONTRIBUTING_GO.md]: Let's clarify what this *IS* then:
```suggestion
2. **Don't break userspace.** This isn't "don't break protocols"; we do, in fact, occasionally break protocols (with care) when we know that users won't notice it. This means support our users' varieties of needs - past, present, and future!
```

File: CONTRIBUTING_GO.md [hunk 3]
```diff
@@ -161,6 +164,27 @@ reasons why this is _the right thing to do_:
...
+3. **Take ownership.** If you're working on something, own it and see it through. Given that (a) we don't have managers and (b) we don't work synchronously, coordination can be difficult. To make up for this, if you want something to happen, you have to make it happen (that may mean calling a meeting, bugging people, fixing unrelated issues, designing alternative solutions, etc.).
+4. **Ask for forgiveness, not permission.** To enable P.3, if you're having trouble getting people to move on something, err on the side of getting things done (bounded by P.1 and P.2). Note: This doesn't apply when you *know* you'll need to ask for forgiveness... use your best judgment.
+
+## Merge Policy
+
+The go merge policy must balance "not breaking things" with "getting things done" and should be read in that spirit.
+
+1. Except in special cases (and mass gx-workspace updates), all changes go through PRs.
+2. Small changes should get one review from someone who can sanity check the change.
```
Comment (momack2) [inline CONTRIBUTING_GO.md]: before being merged by whom?

File: CONTRIBUTING_GO.md [hunk 4]
```diff
@@ -161,6 +164,27 @@ reasons why this is _the right thing to do_:
...
+3. Large changes, especially interface changes, should get signoffs from anyone with a "stake". This is *usually* just the reviewers that GitHub recommends. However, if you're not sure, let someone else merge.
```
Comment (momack2) [inline CONTRIBUTING_GO.md]: Should we encourage people to post a design for large changes somewhere first (like in issues or notes) before they implement it?

File: CONTRIBUTING_GO.md (hunk 2 above)
Comment (Stebalien) [reply CONTRIBUTING_GO.md]: I'm going more for the Linux version of "don't break userspace". That is:

1. Don't change stable APIs that people are relying on.
2. Don't break workflows (within reason).
3. Don't break compatibility (within reason).

File: CONTRIBUTING_GO.md (hunk 3 above)
Comment (Stebalien) [reply CONTRIBUTING_GO.md]: Hm. A merge policy that doesn't state who can merge what...

(the answer is: the PR author merges unless they're unsure, I'll update the doc).

File: CONTRIBUTING_GO.md (hunk 4 above)
Comment (Stebalien) [reply CONTRIBUTING_GO.md]: Yeah, we should, but I'm not sure if that goes *here*. This section is really just about "when can I merge my change".

File: CONTRIBUTING_GO.md (hunk 2 above)
Comment (Stebalien) [reply CONTRIBUTING_GO.md]: I guess this point is really about building trust.

File: CONTRIBUTING_GO.md [hunk 5]
```diff
@@ -161,6 +164,27 @@ reasons why this is _the right thing to do_:
...
```
Comment (eingenito) [inline CONTRIBUTING_GO.md]: Is there a guideline for how many approvals are appropriate before a merge? Is it two always? Or one in sub-projects and two in go-ipfs?

File: CONTRIBUTING_GO.md [hunk 6]
```diff
@@ -161,6 +164,27 @@ reasons why this is _the right thing to do_:
...
+These principles are a living document and should be updated as-needed to better serve the go-ipfs community.
+
+In order of precedence:
+
+1. **Build the distributed peer-to-peer web.** This is our objective and aim - everything we do is in service of this goal and we only compromise our other principles if needed to achieve this.
+2. **Don't break userspace.** This isn't "don't break protocols"; we do, in fact, occasionally break protocols (with care) when we know that users won't notice it.
+3. **Take ownership.** If you're working on something, own it and see it through. Given that (a) we don't have managers and (b) we don't work synchronously, coordination can be difficult. To make up for this, if you want something to happen, you have to make it happen (that may mean calling a meeting, bugging people, fixing unrelated issues, designing alternative solutions, etc.).
+4. **Ask for forgiveness, not permission.** To enable P.3, if you're having trouble getting people to move on something, err on the side of getting things done (bounded by P.1 and P.2). Note: This doesn't apply when you *know* you'll need to ask for forgiveness... use your best judgment.
+
+## Merge Policy
```
Comment (eingenito) [inline CONTRIBUTING_GO.md]: Should this follow the Pull Request Reviews section? There's no reason it has to, in my mind it's probably more important information, but IRL it happens chronologically after a PR is created and prepared for review by others.

**Comment by eingenito** [comment]
I think these changes are great. Thanks for adding this content - the merge policy has always been a little unclear in my mind.

File: CONTRIBUTING_GO.md (hunk 2 above)
Comment (momack2) [reply CONTRIBUTING_GO.md]: Cool - let's add that then?

Comment (terichadbourne) [reply CONTRIBUTING_GO.md]: FWIW, I've never heard the term "userspace" before. Is there a way to describe this concept with different words that are more accessible? From context I guessed it meant "Don't break the user experience" or "Don't make changes that will break apps people already built on top of our code," but Google says it's two words (user space) and means "code that runs outside the operating system's kernel."

Comment (mikeal) [reply CONTRIBUTING_GO.md]: I've seen "userspace" and "ecosystem" used both as differentiated terms and as essentially the same term, so their use without some qualifier is a bit confusing.

Often, "ecosystem" refers to the things built on top of the API/platform directly and "userspace" refers to the users and applications built on that ecosystem.

In Linux, they'll break kernel modules with the understanding that those modules will be updated before release and therefor will not break "userspace" but **will** break ecosystem.

In Node.js, the project can't really break the ecosystem because old packages that depend on those APIs will live in on as deep dependencies in the userspace indefinitely.

Basically, you're allowed to break things provided that you can bring along enough of the ecosystem with you in a reasonable time frame. The way to differentiate "acceptable breakage" from "unacceptable breakage" is the degree to which you think the userspace will be effected.

Comment (Stebalien) [reply CONTRIBUTING_GO.md]: > FWIW, I've never heard the term "userspace" before. Is there a way to describe this concept with different words that are more accessible? From context I guessed it meant "Don't break the user experience" or "Don't make changes that will break apps people already built on top of our code," but Google says it's two words (user space) and means "code that runs outside the operating system's kernel."

This comes from the Linux kernel development and yes, "userspace" isn't really the right term because we're not building a kernel. However, I'd like to find some way to draw the parallel to the Linux kernel because many go-ipfs contributors will be familiar with that phrase and what it entails.

> In Node.js, the project can't really break the ecosystem because old packages that depend on those APIs will live in on as deep dependencies in the userspace indefinitely.

This is more what we're looking for here, I'll try to find a nice way to put this.

File: CONTRIBUTING_GO.md (hunk 5 above)
Comment (Stebalien) [reply CONTRIBUTING_GO.md]: I'm really not a fan of "you need at least X reviewers rules". That has two failure modes:

1. Changes don't get merged because there aren't enough reviews, even if everyone looking at the code *knows* that it's good.
2. Changes get merged when they shouldn't be merged because they check all the right boxes (have tests, have N reviews, etc.).

Solution: If you're not sure if something should be merged, ask around. Ideally, we'd clearly signal who knows what code but we don't have a maintainer/code-owner system in place.

File: CONTRIBUTING_GO.md (hunk 6 above)
Comment (Stebalien) [reply CONTRIBUTING_GO.md]: Maybe? I put this here to motivate that section.

File: CONTRIBUTING_GO.md (hunk 5 above)
Comment (mikeal) [reply CONTRIBUTING_GO.md]: One rule that worked well for Node.js was that, anything could be merged after sufficient time for people to review (48-72 hours) if there weren't objections. Rather than ask for permission to merge the lack of objections was the signal. However, nobody can merge their own PR, so at least one other person has to be willing to get behind it :)

**Comment by Stebalien** [comment]
I've done an initial revision. Please re-read.

File: CONTRIBUTING_GO.md [hunk 7]
```diff
@@ -161,6 +164,45 @@ reasons why this is _the right thing to do_:
...
+* The repo's maintainer (defacto or otherwise).
+* The PR author (if they have write access).
+* A PR reviewer (someone who has reviewed the *entire* PR).
+
+In general, a PR should be merged by someone that knows it's ready and *should not* be merged by someone just because they see that it has been approved. However, use your discretion: reviewers (with write access) should feel free to merge small bug fixes, etc. immediately.
+
+**When can a PR be merged?**
+
+* Small changes should get one review from someone (preferably on the go team if it's a code change) who can sanity check the change.
+* Large changes, especially interface changes, should get signoffs from anyone with a "stake". This is *usually* just the reviewers that GitHub recommends. However, if you're not sure, say so and let someone else merge perform the actual merge. Never be afraid to ask for guidance.
```
Comment (momack2) [inline CONTRIBUTING_GO.md]:
```suggestion
* Large changes, especially interface changes, should get signoffs from anyone with a "stake". This is *usually* just the reviewers that GitHub recommends. However, if you're not sure, say so and let someone else perform the actual merge. Never be afraid to ask for guidance.
```

File: CONTRIBUTING_GO.md [hunk 8]
```diff
@@ -161,6 +164,45 @@ reasons why this is _the right thing to do_:
...
+* All merges to go-ipfs are gated by the maintainer in service of P.2.
```
Comment (momack2) [inline CONTRIBUTING_GO.md]: I'm confused - you give two other ways for merges to be performed, and then you say they're all gated by the maintainer. When/how does that happen if other reviewers are encouraged to merge if they have write access?

File: CONTRIBUTING_GO.md [hunk 9]
```diff
@@ -161,6 +164,45 @@ reasons why this is _the right thing to do_:
...
+## Code Change Policy
+
+The go code change policy must balance "not breaking things" with "getting things done" and should be read in that spirit. Use your best judgment when making changes and PRs; these rules are not set in stone.
+
+**How can the codebase be changed?**
+
+Except in special cases (and mass gx-workspace updates), all changes go through PRs.
+
+* Larger changes should be discussed in an issue *before* work is started to avoid wasted effort, tears, and dead fairies.
+* WIP (work in progress), RFC (request for comments) changes should be marked as such. Not doing so can rouse the review gremlins.
```
Comment (momack2) [inline CONTRIBUTING_GO.md]: cute - but IDK what the review gremlins are. Do you just mean someone might try to review your RFC before you intend them to? OR that it would make people mad / confused if it's an RFC but you don't explicitly say so?

Comment (Stebalien) [reply CONTRIBUTING_GO.md]: Basically, someone will try to review it as if it's a finished product. I'll find something more descriptive.

File: CONTRIBUTING_GO.md (hunk 8 above)
Comment (Stebalien) [reply CONTRIBUTING_GO.md]: To go-ipfs and only go-ipfs. I should probably say "to go-ipfs *itself*".

Comment (momack2) [reply CONTRIBUTING_GO.md]: Still doesn't clarify to me whether as PR author I can actually merge small changes, or as a reviewer I can merge large changes - if the lead maintainer is the gatekeeper. Do you just mean that the maintainer can revert any merged changes, or that the go-ipfs maintainer is the final approver on any/all changes?

Comment (Stebalien) [reply CONTRIBUTING_GO.md]: Yeah, that's still really confusing. Does my latest change clarify this?

Comment (momack2) [reply CONTRIBUTING_GO.md]: MUCH CLEARER! Thank you.

File: CONTRIBUTING_GO.md [hunk 10]
```diff
@@ -161,6 +164,47 @@ reasons why this is _the right thing to do_:
...
+
+## Code Change Policy
+
+The go code change policy must balance "not breaking things" with "getting things done" and should be read in that spirit. Use your best judgment when making changes and PRs; these rules are not set in stone.
+
+**How can the codebase be changed?**
+
+Except in special cases (and mass gx-workspace updates), all changes go through PRs.
+
+* Larger changes should be discussed in an issue *before* work is started to avoid wasted effort, tears, and dead fairies.
```
Comment (djdv) [inline CONTRIBUTING_GO.md]:
```suggestion
* Larger changes should be discussed in an issue *before* work is started, to to avoid wasted effort, and prevent grief.
```
Something like this?
Maybe, sans, commas, though.

File: CONTRIBUTING_GO.md [hunk 11]
```diff
@@ -161,6 +164,47 @@ reasons why this is _the right thing to do_:
...
+* WIP (work in progress), RFC (request for comments) changes should be marked as such. Not doing so can rouse the review gremlins.
```
Comment (djdv) [inline CONTRIBUTING_GO.md]:
```suggestion
* WIP (work in progress), RFC (request for comments) changes should be marked as such. Not doing so is inconsiderate to reviewers, who may think your PR is ready for review. Consider that they may spend time reviewing it, before it is actually ready, so it's best to inform everyone that something is not ready.
```
Okay this isn't concise, but I think it conveys the intention. Maybe it's helpful.

Comment (Stebalien) [reply CONTRIBUTING_GO.md]: I've simplified your change a bit and made it clear that it's not exactly inconsiderate, just frustrating (PR authors are *usually* trying to help).

File: CONTRIBUTING_GO.md (hunk 10 above)
Comment (Stebalien) [reply CONTRIBUTING_GO.md]: I've simplified it to "avoid wasting time".

File: CONTRIBUTING_GO.md [hunk 12]
```diff
@@ -161,6 +164,47 @@ reasons why this is _the right thing to do_:
...
+3. **Take ownership.** If you're working on something, own it and see it through. Given that (a) we don't have managers and (b) we don't work synchronously, coordination can be difficult. To make up for this, if you want something to happen, you have to make it happen (that may mean calling a meeting, bugging people, fixing unrelated issues, designing alternative solutions, etc.).
+4. **Ask for forgiveness, not permission.** To enable P.3, if you're having trouble getting people to move on something, err on the side of getting things done (bounded by P.1 and P.2). Note: This doesn't apply when you *know* you'll need to ask for forgiveness... use your best judgment.
+
+## Code Change Policy
+
+The go code change policy must balance "not breaking things" with "getting things done" and should be read in that spirit. Use your best judgment when making changes and PRs; these rules are not set in stone.
+
+**How can the codebase be changed?**
+
+Except in special cases (and mass gx-workspace updates), all changes go through PRs.
```
Comment (hsanjuan) [inline CONTRIBUTING_GO.md]: It's probably not good to allow people in general to Gx-update directly in master (or do any depedency change directly to master). Right now it is very easy to screw dependencies if you're not an expert. I suggest we ask that all changes go through PRs (including Gx/deps updates), and that in the general case no PR should be merged with failing CI (or at least without verifying and explaining why CI is red). 

(I understand some of us have the experience to bend this rule in some occasions because we can identify risks better and we know how to find and fix our screw ups, but I would not generalize it).

Comment (Stebalien) [reply CONTRIBUTING_GO.md]: You're probably right. And really, it's not *that* hard to do a PR. I just really wish there were a nice way to query PR status from the command line.

Comment (Stebalien) [reply CONTRIBUTING_GO.md]: CI status, really (although I do run the tests locally so it's not *that* much of an issue).

Comment (magik6k) [reply CONTRIBUTING_GO.md]: `hub ci-status -v`?

Comment (Stebalien) [reply CONTRIBUTING_GO.md]: Ok, I *really* need to read that man page.

**Comment by Stebalien** [comment]
@hsanjuan I've addressed your comment.

---

Is someone willing to give this a :heavy_check_mark:?
